File: paimon-core/src/test/java/org/apache/paimon/table/system/SchemasTableTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.table.system;

import org.apache.paimon.CoreOptions;
import org.apache.paimon.catalog.Identifier;
import org.apache.paimon.data.BinaryString;
import org.apache.paimon.data.GenericRow;
import org.apache.paimon.data.InternalRow;
import org.apache.paimon.data.Timestamp;
import org.apache.paimon.data.serializer.InternalRowSerializer;
import org.apache.paimon.fs.FileIO;
import org.apache.paimon.fs.Path;
import org.apache.paimon.fs.local.LocalFileIO;
import org.apache.paimon.predicate.Predicate;
import org.apache.paimon.predicate.PredicateBuilder;
import org.apache.paimon.reader.RecordReader;
import org.apache.paimon.schema.Schema;
import org.apache.paimon.schema.SchemaChange;
import org.apache.paimon.schema.SchemaManager;
import org.apache.paimon.schema.TableSchema;
import org.apache.paimon.table.TableTestBase;
import org.apache.paimon.table.source.ReadBuilder;
import org.apache.paimon.types.DataTypes;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.apache.paimon.utils.JsonSerdeUtil.toFlatJson;
import static org.assertj.core.api.Assertions.assertThat;

/** Unit tests for {@link SchemasTable}. */
public class SchemasTableTest extends TableTestBase {

    private SchemasTable schemasTable;
    private SchemaManager schemaManager;

    @BeforeEach
    public void before() throws Exception {
        Identifier identifier = identifier("T");
        Schema schema =
                Schema.newBuilder()
                        .column("pk", DataTypes.INT())
                        .column("pt", DataTypes.INT())
                        .column("col1", DataTypes.INT())
                        .partitionKeys("pt")
                        .primaryKey("pk", "pt")
                        .option(CoreOptions.CHANGELOG_PRODUCER.key(), "input")
                        .build();
        catalog.createTable(identifier, schema, true);
        schemasTable = (SchemasTable) catalog.getTable(identifier("T$schemas"));

        FileIO fileIO = LocalFileIO.create();
        Path tablePath = new Path(String.format("%s/%s.db/%s", warehouse, database, "T"));
        schemaManager = new SchemaManager(fileIO, tablePath);
    }

    @Test
    public void testSchemasTable() throws Exception {
        List<InternalRow> expectRow = getExpectedResult();
        List<InternalRow> result = read(schemasTable);
        assertThat(result).containsExactlyElementsOf(expectRow);
    }

    @Test
    public void testReadSchemasWithInFilterContainingUnknownId() throws Exception {
        PredicateBuilder builder = new PredicateBuilder(schemasTable.rowType());
        Predicate predicate =
                builder.in(
                        schemasTable.rowType().getFieldNames().indexOf("schema_id"),
                        Arrays.asList(0L, 99L));

        ReadBuilder readBuilder = schemasTable.newReadBuilder().withFilter(predicate);
        List<InternalRow> result = new ArrayList<>();
        InternalRowSerializer serializer = new InternalRowSerializer(schemasTable.rowType());
        readBuilder
                .newRead()
                .createReader(readBuilder.newScan().plan())
                .forEachRemaining(row -> result.add(serializer.copy(row)));

        assertThat(result).containsExactlyElementsOf(getExpectedResult());
    }

    @Test
    public void testReadSchemasWithInAndRangeFilter() throws Exception {
        PredicateBuilder builder = new PredicateBuilder(schemasTable.rowType());
        List<Predicate> predicates =
                Arrays.asList(
                        PredicateBuilder.and(
                                builder.in(0, Arrays.asList(0L, 99L)),
                                builder.greaterOrEqual(0, 0L)),
                        PredicateBuilder.and(
                                builder.in(0, Collections.singletonList(0L)),
                                builder.greaterOrEqual(0, 1L)));
        for (int i = 0; i < predicates.size(); i++) {
            Predicate predicate = predicates.get(i);
            ReadBuilder readBuilder = schemasTable.newReadBuilder().withFilter(predicate);
            List<InternalRow> result = new ArrayList<>();
            InternalRowSerializer serializer = new InternalRowSerializer(schemasTable.rowType());
            readBuilder
                    .newRead()
                    .createReader(readBuilder.newScan().plan())
                    .forEachRemaining(row -> result.add(serializer.copy(row)));

            if (i == 0) {
                assertThat(result).extracting(row -> row.getLong(0)).containsExactly(0L);
            } else {
                assertThat(result).isEmpty();
            }
        }
    }

    @Test
    public void testReadSchemasWithEqualFilterOnUnknownId() throws Exception {
        PredicateBuilder builder = new PredicateBuilder(schemasTable.rowType());
        Predicate predicate =
                builder.equal(schemasTable.rowType().getFieldNames().indexOf("schema_id"), 99L);

        ReadBuilder readBuilder = schemasTable.newReadBuilder().withFilter(predicate);
        List<InternalRow> result = new ArrayList<>();
        readBuilder
                .newRead()
                .createReader(readBuilder.newScan().plan())
                .forEachRemaining(result::add);

        assertThat(result).isEmpty();
    }

    @Test
    public void testReadSchemasWithNestedAndFilter() throws Exception {
        PredicateBuilder builder = new PredicateBuilder(schemasTable.rowType());
        Predicate predicate =
                PredicateBuilder.and(
                        builder.greaterOrEqual(0, 0L),
                        builder.equal(0, 99L),
                        builder.lessThan(0, 3L));

        ReadBuilder readBuilder = schemasTable.newReadBuilder().withFilter(predicate);
        List<InternalRow> result = new ArrayList<>();
        readBuilder
                .newRead()
                .createReader(readBuilder.newScan().plan())
                .forEachRemaining(result::add);

        assertThat(result).isEmpty();
    }

    @Test
    public void testReadSchemasWithExclusiveBoundOverflow() throws Exception {
        PredicateBuilder builder = new PredicateBuilder(schemasTable.rowType());
        for (Predicate predicate :
                Arrays.asList(
                        builder.greaterThan(0, Long.MAX_VALUE),
                        builder.lessThan(0, Long.MIN_VALUE))) {
            ReadBuilder readBuilder = schemasTable.newReadBuilder().withFilter(predicate);
            List<InternalRow> result = new ArrayList<>();
            readBuilder
                    .newRead()
                    .createReader(readBuilder.newScan().plan())
                    .forEachRemaining(result::add);

            assertThat(result).isEmpty();
        }
    }

    @Test
    public void testFilterBySchemaIdEqualAndGreaterOrEqual() throws Exception {
        catalog.alterTable(
                identifier("T"),
                Collections.singletonList(SchemaChange.addColumn("col2", DataTypes.INT())),
                false);
        schemasTable = (SchemasTable) catalog.getTable(identifier("T$schemas"));

        PredicateBuilder builder = new PredicateBuilder(schemasTable.rowType());
        Predicate predicate =
                PredicateBuilder.and(builder.equal(0, 1L), builder.greaterOrEqual(0, 0L));
        ReadBuilder readBuilder = schemasTable.newReadBuilder().withFilter(predicate);
        RecordReader<InternalRow> reader =
                readBuilder.newRead().createReader(readBuilder.newScan().plan());
        List<InternalRow> result = new ArrayList<>();
        InternalRowSerializer serializer = new InternalRowSerializer(schemasTable.rowType());
        reader.forEachRemaining(row -> result.add(serializer.copy(row)));

        assertThat(result).extracting(row -> row.getLong(0)).containsExactly(1L);
    }

    @Test
    public void testFilterBySchemaIdEqualAndLessOrEqual() throws Exception {
        catalog.alterTable(
                identifier("T"),
                Collections.singletonList(SchemaChange.addColumn("col2", DataTypes.INT())),
                false);
        schemasTable = (SchemasTable) catalog.getTable(identifier("T$schemas"));

        PredicateBuilder builder = new PredicateBuilder(schemasTable.rowType());
        for (Predicate predicate :
                Arrays.asList(
                        PredicateBuilder.and(builder.equal(0, 0L), builder.lessOrEqual(0, 1L)),
                        PredicateBuilder.and(builder.lessOrEqual(0, 1L), builder.equal(0, 0L)))) {
            ReadBuilder readBuilder = schemasTable.newReadBuilder().withFilter(predicate);
            RecordReader<InternalRow> reader =
                    readBuilder.newRead().createReader(readBuilder.newScan().plan());
            List<InternalRow> result = new ArrayList<>();
            InternalRowSerializer serializer = new InternalRowSerializer(schemasTable.rowType());
            reader.forEachRemaining(row -> result.add(serializer.copy(row)));

            assertThat(result).extracting(row -> row.getLong(0)).containsExactly(0L);
        }
    }

    @Test
    public void testFilterBySchemaIdWithEmptyRange() throws Exception {
        PredicateBuilder builder = new PredicateBuilder(schemasTable.rowType());
        for (Predicate predicate :
                Arrays.asList(
                        PredicateBuilder.and(builder.equal(0, 0L), builder.greaterOrEqual(0, 1L)),
                        PredicateBuilder.and(
                                builder.greaterOrEqual(0, 1L), builder.equal(0, 0L)))) {
            ReadBuilder readBuilder = schemasTable.newReadBuilder().withFilter(predicate);
            RecordReader<InternalRow> reader =
                    readBuilder.newRead().createReader(readBuilder.newScan().plan());
            List<InternalRow> result = new ArrayList<>();
            reader.forEachRemaining(result::add);

            assertThat(result).isEmpty();
        }
    }

    private List<InternalRow> getExpectedResult() {
        List<TableSchema> tableSchemas = schemaManager.listAll();

        List<InternalRow> expectedRow = new ArrayList<>();
        for (TableSchema schema : tableSchemas) {
            expectedRow.add(
                    GenericRow.of(
                            schema.id(),
                            BinaryString.fromString(toFlatJson(schema.fields())),
                            BinaryString.fromString(toFlatJson(schema.partitionKeys())),
                            BinaryString.fromString(toFlatJson(schema.primaryKeys())),
                            BinaryString.fromString(toFlatJson(schema.options())),
                            BinaryString.fromString(schema.comment()),
                            Timestamp.fromLocalDateTime(
                                    LocalDateTime.ofInstant(
                                            Instant.ofEpochMilli(schema.timeMillis()),
                                            ZoneId.systemDefault()))));
        }
        return expectedRow;
    }
}

File: paimon-python/pypaimon/tests/system/schemas_table_test.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

"""End-to-end tests for the ``$schemas`` system table."""

import datetime
import json
import os
import shutil
import tempfile
import unittest

from pypaimon import CatalogFactory, Schema
from pypaimon.schema.data_types import DataField
from pypaimon.schema.schema_change import SchemaChange
from pypaimon.table.system.schemas_table import SchemasTable


def _read(table):
    rb = table.new_read_builder()
    return rb.new_read().to_arrow(rb.new_scan().plan().splits())


class SchemasTableTest(unittest.TestCase):

    def setUp(self):
        self.tmp = tempfile.mkdtemp(prefix="schemas_sys_")
        warehouse = os.path.join(self.tmp, "warehouse")
        self.catalog = CatalogFactory.create({"warehouse": warehouse})
        self.catalog.create_database("db", False)
        fields = [
            DataField.from_dict({"id": 0, "name": "id", "type": "INT"}),
            DataField.from_dict({"id": 1, "name": "name", "type": "STRING"}),
        ]
        self.catalog.create_table(
            "db.t",
            Schema(fields=fields,
                   options={"user-flag": "alpha"},
                   comment="initial table"),
            False,
        )

    def tearDown(self):
        shutil.rmtree(self.tmp, ignore_errors=True)

    def test_schemas_table_loaded_via_catalog(self):
        table = self.catalog.get_table("db.t$schemas")
        self.assertIsInstance(table, SchemasTable)

    def test_schema_column_layout(self):
        table = self.catalog.get_table("db.t$schemas")
        row_type = table.row_type()
        self.assertEqual(
            ["schema_id", "fields", "partition_keys", "primary_keys",
             "options", "comment", "update_time"],
            [f.name for f in row_type.fields],
        )
        expected_nullability = [False, False, False, False, False, True, False]
        for field, expected in zip(row_type.fields, expected_nullability):
            self.assertEqual(expected, field.type.nullable,
                             "field {} nullability".format(field.name))
        self.assertEqual(["schema_id"], table.primary_keys())

    def test_lists_every_committed_schema_in_id_order(self):
        # Mutate the schema a couple of times so list_all() returns more
        # than one row.
        table = self.catalog.get_table("db.t")
        table.schema_manager.commit_changes(
            [SchemaChange.set_option("user-flag", "beta")])
        table.schema_manager.commit_changes(
            [SchemaChange.set_option("user-flag", "gamma")])

        arrow_table = _read(self.catalog.get_table("db.t$schemas"))
        self.assertEqual([0, 1, 2],
                         arrow_table.column("schema_id").to_pylist())

        fields_json = arrow_table.column("fields").to_pylist()[0]
        decoded = json.loads(fields_json)
        self.assertEqual(["id", "name"], [f["name"] for f in decoded])

        options_jsons = arrow_table.column("options").to_pylist()
        self.assertEqual("alpha", json.loads(options_jsons[0])["user-flag"])
        self.assertEqual("beta", json.loads(options_jsons[1])["user-flag"])
        self.assertEqual("gamma", json.loads(options_jsons[2])["user-flag"])

        comments = arrow_table.column("comment").to_pylist()
        self.assertEqual("initial table", comments[0])

        for ts in arrow_table.column("update_time").to_pylist():
            self.assertIsInstance(ts, datetime.datetime)

    def test_empty_arrays_serialize_to_json_lists(self):
        arrow_table = _read(self.catalog.get_table("db.t$schemas"))
        self.assertEqual("[]",
                         arrow_table.column("partition_keys").to_pylist()[0])
        self.assertEqual("[]",
                         arrow_table.column("primary_keys").to_pylist()[0])


if __name__ == "__main__":
    unittest.main()
